File: MakEliz/backend/routes/manager.py
# ============================================================
# manager.py
# Maintenance Manager routes
# - View EM requests sorted by priority
# - View all Accountable Parties
# - Assign assets to Accountable Parties
# - View work received forms
# - View work completion forms submitted by Accountable Parties
# ============================================================

from flask import Blueprint, request, jsonify, session
import sys
import os

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from database import get_connection

manager_bp = Blueprint("manager", __name__)


def require_manager():
    if "employee_id" not in session:
        return False
    if session.get("role") not in ["Maintenance Manager", "Admin"]:
        return False
    return True


# ============================================================
# GET: All open EM requests sorted by priority
# ============================================================

@manager_bp.route("/em-requests", methods=["GET"])
def get_em_requests():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                a.AssetSN,
                a.AssetName,
                em.EMReportDate,
                e.FirstName + ' ' + e.LastName AS EmployeeFullName,
                d.Name AS Department,
                p.Name AS Priority,
                em.ID AS EmergencyID
            FROM EmergencyMaintenances em
            JOIN Assets a ON em.AssetID = a.ID
            JOIN Employees e ON a.EmployeeID = e.ID
            JOIN DepartmentLocations dl ON a.DepartmentLocationID = dl.ID
            JOIN Departments d ON dl.DepartmentID = d.ID
            JOIN Priorities p ON em.PriorityID = p.ID
            WHERE em.EMEndDate IS NULL
            ORDER BY
                CASE p.Name
                    WHEN 'Very High' THEN 1
                    WHEN 'High' THEN 2
                    WHEN 'Normal' THEN 3
                    ELSE 4
                END
            """
        )
        rows = cursor.fetchall()
        conn.close()

        requests = []
        for row in rows:
            requests.append({
                "asset_sn": row[0],
                "asset_name": row[1],
                "request_date": str(row[2]) if row[2] else "",
                "employee_full_name": row[3],
                "department": row[4],
                "priority": row[5],
                "emergency_id": row[6]
            })

        return jsonify({"success": True, "requests": requests})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500


# ============================================================
# GET: All assets in the system (for the assign form dropdown)
# ============================================================

@manager_bp.route("/all-assets", methods=["GET"])
def get_all_assets():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                a.ID,
                a.AssetSN,
                a.AssetName,
                d.Name AS Department
            FROM Assets a
            LEFT JOIN DepartmentLocations dl ON a.DepartmentLocationID = dl.ID
            LEFT JOIN Departments d ON dl.DepartmentID = d.ID
            ORDER BY a.AssetName
            """
        )
        rows = cursor.fetchall()
        conn.close()

        assets = []
        for row in rows:
            assets.append({
                "id": row[0],
                "asset_sn": row[1],
                "asset_name": row[2],
                "department": row[3] or ""
            })

        return jsonify({"success": True, "assets": assets})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500


# ============================================================
# GET: All registered Accountable Parties
# ============================================================

@manager_bp.route("/accountable-parties", methods=["GET"])
def get_accountable_parties():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT ID, FirstName, LastName, Email, Phone
            FROM Employees
            WHERE Role = 'Accountable Party'
            ORDER BY FirstName
            """
        )
        rows = cursor.fetchall()
        conn.close()

        parties = []
        for row in rows:
            parties.append({
                "id": row[0],
                "first_name": row[1],
                "last_name": row[2],
                "email": row[3],
                "phone": row[4] or ""
            })

        return jsonify({"success": True, "parties": parties})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500


# ============================================================
# POST: Assign an asset to an Accountable Party
# ============================================================

@manager_bp.route("/assign-asset", methods=["POST"])
def assign_asset():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    data = request.get_json()
    asset_id = data.get("asset_id")
    assigned_to = data.get("assigned_to_employee_id")
    assigned_date = data.get("assigned_date")
    notes = data.get("notes", "")

    if not asset_id or not assigned_to or not assigned_date:
        return jsonify({"success": False, "message": "Asset, employee, and date are required."}), 400

    manager_id = session.get("employee_id")

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            INSERT INTO AssetAssignments (AssetID, AssignedToEmployeeID, AssignedByEmployeeID, AssignedDate, Notes, Status)
            VALUES (?, ?, ?, ?, ?, 'Assigned')
            """,
            (asset_id, assigned_to, manager_id, assigned_date, notes)
        )
        conn.commit()
        conn.close()

        return jsonify({"success": True, "message": "Asset assigned successfully."})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500


# ============================================================
# GET: All assignments made by this manager
# ============================================================

@manager_bp.route("/assignments", methods=["GET"])
def get_assignments():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                aa.ID,
                a.AssetSN,
                a.AssetName,
                emp.FirstName + ' ' + emp.LastName AS AssignedTo,
                emp.Email AS AssignedToEmail,
                aa.AssignedDate,
                aa.Notes,
                aa.Status
            FROM AssetAssignments aa
            JOIN Assets a ON aa.AssetID = a.ID
            JOIN Employees emp ON aa.AssignedToEmployeeID = emp.ID
            ORDER BY aa.AssignedDate DESC
            """
        )
        rows = cursor.fetchall()
        conn.close()

        assignments = []
        for row in rows:
            assignments.append({
                "id": row[0],
                "asset_sn": row[1],
                "asset_name": row[2],
                "assigned_to": row[3],
                "assigned_to_email": row[4],
                "assigned_date": str(row[5]) if row[5] else "",
                "notes": row[6] or "",
                "status": row[7]
            })

        return jsonify({"success": True, "assignments": assignments})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500


# ============================================================
# GET: All work completion forms submitted to this manager
# ============================================================

@manager_bp.route("/completion-forms", methods=["GET"])
def get_completion_forms():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                wc.ID,
                a.AssetSN,
                a.AssetName,
                emp.FirstName + ' ' + emp.LastName AS SubmittedBy,
                wc.CompletionDate,
                wc.WorkDoneDescription,
                wc.ProblemsEncountered,
                wc.RecommendedActions,
                wc.FinalCondition,
                aa.Status
            FROM WorkCompletionForms wc
            JOIN AssetAssignments aa ON wc.AssignmentID = aa.ID
            JOIN Assets a ON aa.AssetID = a.ID
            JOIN Employees emp ON wc.SubmittedByEmployeeID = emp.ID
            ORDER BY wc.CompletionDate DESC
            """
        )
        rows = cursor.fetchall()
        conn.close()

        forms = []
        for row in rows:
            forms.append({
                "id": row[0],
                "asset_sn": row[1],
                "asset_name": row[2],
                "submitted_by": row[3],
                "completion_date": str(row[4]) if row[4] else "",
                "work_done": row[5] or "",
                "problems": row[6] or "",
                "recommended_actions": row[7] or "",
                "final_condition": row[8] or "",
                "status": row[9]
            })

        return jsonify({"success": True, "forms": forms})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500


# ============================================================
# GET: Work received forms submitted by Accountable Parties
# ============================================================

@manager_bp.route("/received-forms", methods=["GET"])
def get_received_forms():
    if not require_manager():
        return jsonify({"success": False, "message": "Access denied."}), 403

    try:
        conn = get_connection()
        cursor = conn.cursor()

        cursor.execute(
            """
            SELECT
                wr.ID,
                a.AssetSN,
                a.AssetName,
                emp.FirstName + ' ' + emp.LastName AS ReceivedBy,
                wr.ReceivedDate,
                wr.ConditionOnReceiving,
                wr.CommentsOnReceiving
            FROM WorkReceivedForms wr
            JOIN AssetAssignments aa ON wr.AssignmentID = aa.ID
            JOIN Assets a ON aa.AssetID = a.ID
            JOIN Employees emp ON wr.ReceivedByEmployeeID = emp.ID
            ORDER BY wr.ReceivedDate DESC
            """
        )
        rows = cursor.fetchall()
        conn.close()

        forms = []
        for row in rows:
            forms.append({
                "id": row[0],
                "asset_sn": row[1],
                "asset_name": row[2],
                "received_by": row[3],
                "received_date": str(row[4]) if row[4] else "",
                "condition": row[5] or "",
                "comments": row[6] or ""
            })

        return jsonify({"success": True, "forms": forms})

    except Exception as e:
        return jsonify({"success": False, "message": "Database error: " + str(e)}), 500

File: MakEliz/frontend/js/manager.js
// ============================================================
// manager.js
// Handles the Maintenance Manager Dashboard
// ============================================================

var API_BASE = "http://127.0.0.1:5000/api";

// ============================================================
// On page load
// ============================================================

window.onload = function () {
    checkSession();
};

function checkSession() {
    fetch(API_BASE + "/auth/session", {
        method: "GET",
        credentials: "include"
    })
    .then(function (response) { return response.json(); })
    .then(function (data) {
        if (!data.logged_in || data.role !== "Maintenance Manager") {
            window.location.href = "login.html";
        } else {
            document.getElementById("manager-name").textContent = data.name;
            loadEMRequests();
        }
    })
    .catch(function () {
        window.location.href = "login.html";
    });
}

// ============================================================
// Navigation
// ============================================================

function showSection(sectionId, clickedItem) {
    document.querySelectorAll(".page-section").forEach(function (s) {
        s.classList.remove("active");
    });
    document.querySelectorAll(".sidebar-nav li").forEach(function (li) {
        li.classList.remove("active");
    });

    document.getElementById(sectionId).classList.add("active");
    clickedItem.classList.add("active");

    if (sectionId === "section-requests")    { loadEMRequests(); }
    if (sectionId === "section-assign")      { loadAssignDropdowns(); }
    if (sectionId === "section-assignments") { loadAssignments(); }
    if (sectionId === "section-received")    { loadReceivedForms(); }
    if (sectionId === "section-completions") { loadCompletionForms(); }
}

// ============================================================
// Show message helper
// ============================================================

function showMessage(elementId, text, type) {
    var el = document.getElementById(elementId);
    el.className = "message " + type;
    el.textContent = text;
}

// ============================================================
// EM Requests
// ============================================================

function loadEMRequests() {
    var tbody = document.getElementById("requests-table-body");
    tbody.innerHTML = "<tr><td colspan='6' class='no-data'>Loading...</td></tr>";

    fetch(API_BASE + "/manager/em-requests", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (!data.success || data.requests.length === 0) {
            tbody.innerHTML = "<tr><td colspan='6' class='no-data'>No open EM requests.</td></tr>";
            return;
        }

        var html = "";
        data.requests.forEach(function (req) {
            var badge = req.priority === "Very High" ? "badge-veryhigh"
                      : req.priority === "High"      ? "badge-high"
                      : "badge-normal";

            html += "<tr>";
            html += "<td>" + req.asset_sn + "</td>";
            html += "<td>" + req.asset_name + "</td>";
            html += "<td>" + req.request_date + "</td>";
            html += "<td>" + req.employee_full_name + "</td>";
            html += "<td>" + req.department + "</td>";
            html += "<td><span class='badge " + badge + "'>" + req.priority + "</span></td>";
            html += "</tr>";
        });
        tbody.innerHTML = html;
    })
    .catch(function () {
        tbody.innerHTML = "<tr><td colspan='6' class='no-data'>Server error.</td></tr>";
    });
}

// ============================================================
// Assign Asset - load dropdowns
// ============================================================

function loadAssignDropdowns() {
    var assetSelect = document.getElementById("assign-asset");
    var partySelect = document.getElementById("assign-party");

    assetSelect.innerHTML = "<option value=''>Loading assets...</option>";
    partySelect.innerHTML = "<option value=''>Loading parties...</option>";

    fetch(API_BASE + "/manager/all-assets", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        assetSelect.innerHTML = "<option value=''>-- Select Asset --</option>";
        if (data.success) {
            data.assets.forEach(function (asset) {
                assetSelect.innerHTML += "<option value='" + asset.id + "'>"
                    + asset.asset_sn + " - " + asset.asset_name
                    + (asset.department ? " (" + asset.department + ")" : "")
                    + "</option>";
            });
        }
    });

    fetch(API_BASE + "/manager/accountable-parties", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        partySelect.innerHTML = "<option value=''>-- Select Person --</option>";
        if (data.success) {
            data.parties.forEach(function (p) {
                partySelect.innerHTML += "<option value='" + p.id + "'>"
                    + p.first_name + " " + p.last_name + " - " + p.email
                    + "</option>";
            });
        }
    });
}

function submitAssignment() {
    var assetId   = document.getElementById("assign-asset").value;
    var partyId   = document.getElementById("assign-party").value;
    var date      = document.getElementById("assign-date").value;
    var notes     = document.getElementById("assign-notes").value.trim();

    if (!assetId || !partyId || !date) {
        showMessage("assign-message", "Please select an asset, a person, and a date.", "error");
        return;
    }

    fetch(API_BASE + "/manager/assign-asset", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        credentials: "include",
        body: JSON.stringify({
            asset_id: assetId,
            assigned_to_employee_id: partyId,
            assigned_date: date,
            notes: notes
        })
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (data.success) {
            showMessage("assign-message", "Asset assigned successfully.", "success");
            document.getElementById("assign-asset").value = "";
            document.getElementById("assign-party").value = "";
            document.getElementById("assign-date").value = "";
            document.getElementById("assign-notes").value = "";
        } else {
            showMessage("assign-message", data.message, "error");
        }
    })
    .catch(function () {
        showMessage("assign-message", "Server error. Could not assign asset.", "error");
    });
}

// ============================================================
// All Assignments
// ============================================================

function loadAssignments() {
    var tbody = document.getElementById("assignments-table-body");
    tbody.innerHTML = "<tr><td colspan='6' class='no-data'>Loading...</td></tr>";

    fetch(API_BASE + "/manager/assignments", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (!data.success || data.assignments.length === 0) {
            tbody.innerHTML = "<tr><td colspan='6' class='no-data'>No assignments found.</td></tr>";
            return;
        }

        var html = "";
        data.assignments.forEach(function (a) {
            var statusClass = a.status === "Completed"  ? "badge-normal"
                            : a.status === "In Progress" ? "badge-high"
                            : "badge-admin";

            html += "<tr>";
            html += "<td>" + a.asset_sn + "</td>";
            html += "<td>" + a.asset_name + "</td>";
            html += "<td>" + a.assigned_to + "</td>";
            html += "<td>" + a.assigned_date + "</td>";
            html += "<td>" + (a.notes || "-") + "</td>";
            html += "<td><span class='badge " + statusClass + "'>" + a.status + "</span></td>";
            html += "</tr>";
        });
        tbody.innerHTML = html;
    })
    .catch(function () {
        tbody.innerHTML = "<tr><td colspan='6' class='no-data'>Server error.</td></tr>";
    });
}

// ============================================================
// Work Received Forms
// ============================================================

function loadReceivedForms() {
    var tbody = document.getElementById("received-table-body");
    tbody.innerHTML = "<tr><td colspan='6' class='no-data'>Loading...</td></tr>";

    fetch(API_BASE + "/manager/received-forms", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (!data.success || data.forms.length === 0) {
            tbody.innerHTML = "<tr><td colspan='6' class='no-data'>No received forms submitted yet.</td></tr>";
            return;
        }

        var html = "";
        data.forms.forEach(function (f) {
            html += "<tr>";
            html += "<td>" + f.asset_sn + "</td>";
            html += "<td>" + f.asset_name + "</td>";
            html += "<td>" + f.received_by + "</td>";
            html += "<td>" + f.received_date + "</td>";
            html += "<td>" + (f.condition || "-") + "</td>";
            html += "<td>" + (f.comments || "-") + "</td>";
            html += "</tr>";
        });
        tbody.innerHTML = html;
    })
    .catch(function () {
        tbody.innerHTML = "<tr><td colspan='6' class='no-data'>Server error.</td></tr>";
    });
}



function loadCompletionForms() {
    var tbody = document.getElementById("completions-table-body");
    tbody.innerHTML = "<tr><td colspan='8' class='no-data'>Loading...</td></tr>";

    fetch(API_BASE + "/manager/completion-forms", {
        method: "GET",
        credentials: "include"
    })
    .then(function (r) { return r.json(); })
    .then(function (data) {
        if (!data.success || data.forms.length === 0) {
            tbody.innerHTML = "<tr><td colspan='8' class='no-data'>No completion forms submitted yet.</td></tr>";
            return;
        }

        var html = "";
        data.forms.forEach(function (f) {
            html += "<tr>";
            html += "<td>" + f.asset_sn + "</td>";
            html += "<td>" + f.asset_name + "</td>";
            html += "<td>" + f.submitted_by + "</td>";
            html += "<td>" + f.completion_date + "</td>";
            html += "<td>" + (f.work_done || "-") + "</td>";
            html += "<td>" + (f.problems || "-") + "</td>";
            html += "<td>" + (f.recommended_actions || "-") + "</td>";
            html += "<td>" + (f.final_condition || "-") + "</td>";
            html += "</tr>";
        });
        tbody.innerHTML = html;
    })
    .catch(function () {
        tbody.innerHTML = "<tr><td colspan='8' class='no-data'>Server error.</td></tr>";
    });
}

function handleLogout() {
    fetch(API_BASE + "/auth/logout", { method: "POST", credentials: "include" })
    .then(function () { window.location.href = "login.html"; })
    .catch(function () { window.location.href = "login.html"; });
}